diff --git a/Makefile b/Makefile
index 39a99d7..929f6d4 100644
--- a/Makefile
+++ b/Makefile
@@ -132,6 +132,10 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
+	$U/_threads\
+	$U/_test\
+	$U/_mtest\
+
 
 fs.img: mkfs/mkfs README $(UPROGS)
 	mkfs/mkfs fs.img README $(UPROGS)
diff --git a/kernel/defs.h b/kernel/defs.h
index a3c962b..a00a929 100644
--- a/kernel/defs.h
+++ b/kernel/defs.h
@@ -106,6 +106,11 @@ void            yield(void);
 int             either_copyout(int user_dst, uint64 dst, void *src, uint64 len);
 int             either_copyin(void *dst, int user_src, uint64 src, uint64 len);
 void            procdump(void);
+int             create_thread(void (*fcn)(void*) , void* arg , void* stack) ;
+int             join_thread(int id) ;
+void            exit_thread() ;
+void            new_sleep(uint64) ;
+void            new_wakeup(int) ;
 
 // swtch.S
 void            swtch(struct context*, struct context*);
@@ -173,7 +178,10 @@ uint64          walkaddr(pagetable_t, uint64);
 int             copyout(pagetable_t, uint64, char *, uint64);
 int             copyin(pagetable_t, char *, uint64, uint64);
 int             copyinstr(pagetable_t, char *, uint64, uint64);
-
+int             uvmmirror(pagetable_t ,pagetable_t,uint64,uint64) ;
+void            freewalk(pagetable_t) ;
+uint64          uvmallocmirror(pagetable_t ,pagetable_t, uint64 , uint64) ;
+uint64          uvmdeallocmirror(pagetable_t , uint64 , uint64) ;
 // plic.c
 void            plicinit(void);
 void            plicinithart(void);
diff --git a/kernel/proc.c b/kernel/proc.c
index 959b778..771d685 100644
--- a/kernel/proc.c
+++ b/kernel/proc.c
@@ -9,10 +9,11 @@
 struct cpu cpus[NCPU];
 
 struct proc proc[NPROC];
+struct spinlock memlocks[NPROC] ;
 
 struct proc *initproc;
 
-int nextpid = 1;
+int nextpid = 1 ;
 struct spinlock pid_lock;
 
 extern void forkret(void);
@@ -50,9 +51,11 @@ procinit(void)
   struct proc *p;
   
   initlock(&pid_lock, "nextpid");
-  initlock(&wait_lock, "wait_lock");
+  initlock(&wait_lock, "wait_lock") ;
+  int i = 0 ;
   for(p = proc; p < &proc[NPROC]; p++) {
       initlock(&p->lock, "proc");
+      initlock(&memlocks[i++], "mem") ; 
       p->state = UNUSED;
       p->kstack = KSTACK((int) (p - proc));
   }
@@ -102,6 +105,7 @@ allocpid()
   return pid;
 }
 
+
 // Look in the process table for an UNUSED proc.
 // If found, initialize state required to run in the kernel,
 // and return with p->lock held.
@@ -110,20 +114,25 @@ static struct proc*
 allocproc(void)
 {
   struct proc *p;
-
+  int i = 0 ;
   for(p = proc; p < &proc[NPROC]; p++) {
     acquire(&p->lock);
     if(p->state == UNUSED) {
+      p->memlock = &memlocks[i] ;
+      //printf("%d\n",i) ;
       goto found;
     } else {
       release(&p->lock);
     }
+    i++ ;
   }
   return 0;
 
 found:
   p->pid = allocpid();
+  p->mem_id = p->pid ;
   p->state = USED;
+  p->is_thread = 0 ;
 
   // Allocate a trapframe page.
   if((p->trapframe = (struct trapframe *)kalloc()) == 0){
@@ -155,11 +164,23 @@ found:
 static void
 freeproc(struct proc *p)
 {
+  acquire(p->memlock) ;
   if(p->trapframe)
     kfree((void*)p->trapframe);
-  p->trapframe = 0;
-  if(p->pagetable)
-    proc_freepagetable(p->pagetable, p->sz);
+  p->trapframe = 0 ;
+
+  //modifying the proc_freepagetable()
+  if(p->pagetable) {
+    uvmunmap(p->pagetable, TRAMPOLINE, 1, 0) ;
+    uvmunmap(p->pagetable, TRAPFRAME,  1, 0) ;
+    if(p->sz > 0) {
+      // if p->is_thread = 1 then it will not freeing the frames/pages of memory!
+      //because we didnot allocate(call kalloc()) in memory for thread creation
+      uvmunmap(p->pagetable, 0, PGROUNDUP(p->sz)/PGSIZE,!(p->is_thread));
+    }
+    freewalk(p->pagetable);
+  }
+  release(p->memlock) ;
   p->pagetable = 0;
   p->sz = 0;
   p->pid = 0;
@@ -169,6 +190,9 @@ freeproc(struct proc *p)
   p->killed = 0;
   p->xstate = 0;
   p->state = UNUSED;
+  p->is_thread = 0 ;
+  p->mem_id = 0 ;
+  p->memlock = 0 ;
 }
 
 // Create a user page table for a given process, with no user memory,
@@ -262,15 +286,40 @@ growproc(int n)
   uint64 sz;
   struct proc *p = myproc();
 
+  acquire(p->memlock) ;
   sz = p->sz;
+
+  //for normal process / child process
   if(n > 0){
     if((sz = uvmalloc(p->pagetable, sz, sz + n, PTE_W)) == 0) {
+      release(p->memlock) ;
       return -1;
     }
   } else if(n < 0){
     sz = uvmdealloc(p->pagetable, sz, sz + n);
   }
-  p->sz = sz;
+  p->sz = sz ;
+  //release(p->memlock) ;
+
+
+  //newly added for to increase/decrease size in physical memory of all threads pointing to same memory location
+  //acquire(p->memlock) ;
+  for(struct proc* pr = proc ; pr < &proc[NPROC] ; pr++) {
+    acquire(&pr->lock) ;
+    if(p->pid != pr->pid && p->mem_id == pr->mem_id) {
+      sz = pr->sz ;
+      if(n > 0) {
+        //printf("HGSFDG\n\n") ;
+        sz = uvmallocmirror(p->pagetable, pr->pagetable ,sz, sz+n) ;
+        pr->sz = !sz ? pr->sz : sz ;
+      } 
+      else if(n < 0) {
+        pr->sz = uvmdeallocmirror(p->pagetable, sz, sz + n);
+      }
+    }
+    release(&pr->lock) ;
+  }
+  release(p->memlock) ;
   return 0;
 }
 
@@ -287,7 +336,6 @@ fork(void)
   if((np = allocproc()) == 0){
     return -1;
   }
-
   // Copy user memory from parent to child.
   if(uvmcopy(p->pagetable, np->pagetable, p->sz) < 0){
     freeproc(np);
@@ -311,6 +359,7 @@ fork(void)
   safestrcpy(np->name, p->name, sizeof(p->name));
 
   pid = np->pid;
+  np->is_thread = 0 ;
 
   release(&np->lock);
 
@@ -681,3 +730,138 @@ procdump(void)
     printf("\n");
   }
 }
+
+int
+create_thread(void(*fcn)(void*) , void* arg , void* stack) {
+  int i, pid;
+  struct proc *np;
+  struct proc *p = myproc();
+
+  // Allocate process.
+  if((np = allocproc()) == 0){
+    return -1;
+  }
+
+  //just use the same adress space, same content
+  //donot allocate memory
+  if(uvmmirror(p->pagetable, np->pagetable, (uint64) 0, p->sz) < 0){
+    freeproc(np);
+    release(&np->lock);
+    return -1;
+  }
+  np->memlock = p->memlock ;
+  acquire(np->memlock) ;
+  np->sz = p->sz ;
+  release(np->memlock) ;
+  np->is_thread = 1 ;
+  np->mem_id = p->mem_id ;
+
+  // copy saved user registers.
+  *(np->trapframe) = *(p->trapframe);
+
+  //taking the arguments in a0 resgister
+  np->trapframe->a0 = (uint64)arg ;
+  np->trapframe->epc = (uint64) fcn ;
+  np->trapframe->sp = (1 << 12) + (uint64) stack ;  
+  np->trapframe->sp-= np->trapframe->sp % 16 ; // same from exec.c
+  np->trapframe->ra = 0xffffffff ;
+
+  // increment reference counts on open file descriptors.
+  for(i = 0; i < NOFILE; i++)
+    if(p->ofile[i])
+      np->ofile[i] = filedup(p->ofile[i]);
+  np->cwd = idup(p->cwd);
+
+  safestrcpy(np->name, p->name, sizeof(p->name));
+
+  pid = np->pid;
+
+  release(&np->lock);
+
+  acquire(&wait_lock);
+  np->parent = p;
+  release(&wait_lock);
+
+  acquire(&np->lock);
+  np->state = RUNNABLE;
+  release(&np->lock);
+
+  return pid;
+}
+
+int 
+join_thread(int id) {
+  struct proc *pp;
+  int havekids, pid;
+  struct proc *p = myproc();
+
+  acquire(&wait_lock);
+
+  for(;;){
+    // Scan through table looking for exited children.
+    havekids = 0;
+    for(pp = proc; pp < &proc[NPROC]; pp++){
+      if(pp->parent == p && pp->pid == id && pp->mem_id == p->mem_id){
+        // make sure the child isn't still in exit() or swtch().
+        acquire(&pp->lock);
+
+        havekids = 1;
+        if(pp->state == ZOMBIE){
+          // Found one.
+          pid = pp->pid ;
+          freeproc(pp);
+          release(&pp->lock);
+          release(&wait_lock);
+          return pid ;
+        }
+        release(&pp->lock);
+      }
+    }
+
+    // No point waiting if we don't have any children.
+    if(!havekids || killed(p)){
+      release(&wait_lock);
+      return -1;
+    }
+    
+    // Wait for a child to exit.
+    sleep(p, &wait_lock);  //DOC: wait-sleep
+  }
+}
+
+void
+exit_thread(void) {
+  exit(0) ;
+}
+
+void new_sleep(uint64 locked) {
+
+    struct proc* p = myproc() ;
+    acquire(&p->lock) ;
+    uint64 va0, pa0;
+    va0 = PGROUNDDOWN(locked);
+    pa0 = walkaddr(myproc()->pagetable, va0);
+    // if(pa0 == 0)
+    //   return -1;
+    __sync_synchronize();
+    __sync_lock_release((uint8*)(pa0 + (locked - va0)));
+
+    p->state = SLEEPING ;
+    sched() ;
+    
+    release(&p->lock) ;
+}
+
+void new_wakeup(int id) {
+  struct proc *p;
+
+  for(p = proc; p < &proc[NPROC]; p++) {
+    if(p != myproc()){
+      acquire(&p->lock);
+      if(p->state == SLEEPING && p->pid == id) {
+        p->state = RUNNABLE;
+      }
+      release(&p->lock);
+    }
+  }
+}
diff --git a/kernel/proc.h b/kernel/proc.h
index d021857..7342f14 100644
--- a/kernel/proc.h
+++ b/kernel/proc.h
@@ -104,4 +104,7 @@ struct proc {
   struct file *ofile[NOFILE];  // Open files
   struct inode *cwd;           // Current directory
   char name[16];               // Process name (debugging)
+  struct spinlock* memlock ;
+  int mem_id ;
+  int is_thread ;
 };
diff --git a/kernel/syscall.c b/kernel/syscall.c
index ed65409..f17c9f0 100644
--- a/kernel/syscall.c
+++ b/kernel/syscall.c
@@ -101,6 +101,12 @@ extern uint64 sys_unlink(void);
 extern uint64 sys_link(void);
 extern uint64 sys_mkdir(void);
 extern uint64 sys_close(void);
+extern uint64 sys_create_thread(void);
+extern uint64 sys_exit_thread(void);
+extern uint64 sys_join_thread(void);
+extern uint64 sys_yield(void) ;
+extern uint64 sys_release_sleep(void) ;
+extern uint64 sys_wakeup(void) ;
 
 // An array mapping syscall numbers from syscall.h
 // to the function that handles the system call.
@@ -126,6 +132,12 @@ static uint64 (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+[SYS_create_thread]   sys_create_thread,
+[SYS_exit_thread]   sys_exit_thread,
+[SYS_join_thread]   sys_join_thread,
+[SYS_yield]   sys_yield ,
+[SYS_release_sleep]   sys_release_sleep ,
+[SYS_wakeup]   sys_wakeup ,
 };
 
 void
diff --git a/kernel/syscall.h b/kernel/syscall.h
index bc5f356..c3a38c9 100644
--- a/kernel/syscall.h
+++ b/kernel/syscall.h
@@ -20,3 +20,10 @@
 #define SYS_link   19
 #define SYS_mkdir  20
 #define SYS_close  21
+#define SYS_create_thread  22
+#define SYS_exit_thread    23
+#define SYS_join_thread    24
+#define SYS_yield 25
+#define SYS_release_sleep 26
+#define SYS_wakeup 27
+
diff --git a/kernel/sysproc.c b/kernel/sysproc.c
index 1de184e..9190a74 100644
--- a/kernel/sysproc.c
+++ b/kernel/sysproc.c
@@ -89,3 +89,53 @@ sys_uptime(void)
   release(&tickslock);
   return xticks;
 }
+
+uint64
+sys_create_thread(void) {
+  uint64 fcn ;
+  uint64 arg ;
+  uint64 stack ;
+  argaddr(0, &fcn) ; 
+  argaddr(1 , &arg) ;
+  argaddr(2, &stack) ;
+
+  return create_thread((void*) fcn, (void*) arg, (void*) stack) ;
+}
+
+uint64
+sys_exit_thread(void) {
+  exit_thread() ;
+  return 0 ;
+}
+
+uint64
+sys_join_thread(void) {
+  int id ;
+  argint(0,&id) ;
+  return join_thread(id) ;
+}
+
+uint64
+sys_yield(void) {
+  int x ;
+  argint(0,&x) ;
+  yield() ;
+  return 0 ;
+}
+
+uint64
+sys_release_sleep(void) {
+  uint64 addr ;
+  argaddr(0, &addr) ;
+  new_sleep(addr) ;
+
+  return 0 ;
+}
+
+uint64
+sys_wakeup(void) {
+  int id ;
+  argint(0, &id) ;
+  new_wakeup(id) ;
+  return 0 ;
+}
\ No newline at end of file
diff --git a/kernel/vm.c b/kernel/vm.c
index 9f69783..21bc49e 100644
--- a/kernel/vm.c
+++ b/kernel/vm.c
@@ -290,9 +290,9 @@ freewalk(pagetable_t pagetable)
 // then free page-table pages.
 void
 uvmfree(pagetable_t pagetable, uint64 sz)
-{
+{ 
   if(sz > 0)
-    uvmunmap(pagetable, 0, PGROUNDUP(sz)/PGSIZE, 1);
+    uvmunmap(pagetable, 0, PGROUNDUP(sz)/PGSIZE,1);
   freewalk(pagetable);
 }
 
@@ -332,6 +332,82 @@ uvmcopy(pagetable_t old, pagetable_t new, uint64 sz)
   return -1;
 }
 
+int
+uvmmirror(pagetable_t old, pagetable_t new, uint64 oldsz , uint64 newsz)
+{
+  pte_t *pte;
+  uint64 pa, i;
+  uint flags;
+
+  //newsz need to be aligned. if oldsz isd 3.5 KB and and newsz is 3.8 KB..round up 
+  //of oldsz makes it 4096. But that size is allocated already using kalloc() before, kalloc() allocates 4KB for each page
+  //but whenever we have newsz = 4.6KB then we need iterate the loop to 4 KB + 4KB block.Because kalloc always allocate 
+  //4KB for each allocation!
+  oldsz = PGROUNDUP(oldsz) ; //if oldsz is 0 the PG-ROUND-UP makes it 0.(0+4095) & (~4095) = 0 
+  for(i = oldsz ; i < newsz; i += PGSIZE){
+    if((pte = walk(old, i, 0)) == 0) 
+      panic("uvmmirror: pte should exist");
+    if((*pte & PTE_V) == 0)
+      panic("uvmmirror: page not present");
+    pa = PTE2PA(*pte);
+    flags = PTE_FLAGS(*pte);
+    if(mappages(new, i, PGSIZE, (uint64)pa, flags) != 0){
+      goto err;
+    }
+  }
+  return newsz ;
+
+ err:
+  uvmunmap(new, 0, i / PGSIZE, 0);
+  return -1;
+}
+
+uint64
+uvmallocmirror(pagetable_t parentptable,pagetable_t threadptable, uint64 oldsz, uint64 newsz)
+{
+  pte_t *pte;
+  uint64 pa, i;
+  uint flags;
+
+
+   if(newsz < oldsz)
+    return oldsz;
+
+  oldsz = PGROUNDUP(oldsz);
+  for(i = oldsz; i < newsz; i += PGSIZE){
+    if((pte = walk(parentptable, i, 0)) == 0)
+      panic("uvmmallocmirror: pte should exist");
+    if((*pte & PTE_V) == 0)
+      panic("uvmmallocmirror: page not present");
+    pa = PTE2PA(*pte);
+    flags = PTE_FLAGS(*pte);
+    if(mappages(threadptable, i, PGSIZE, (uint64)pa, flags) != 0){
+      uvmdeallocmirror(threadptable,i,oldsz);
+      return 0;
+    }
+  }
+  return newsz;
+
+}
+
+uint64
+uvmdeallocmirror(pagetable_t pagetable, uint64 oldsz, uint64 newsz)
+{
+  if(newsz >= oldsz)
+    return oldsz;
+
+  if(PGROUNDUP(newsz) < PGROUNDUP(oldsz)){
+    int npages = (PGROUNDUP(oldsz) - PGROUNDUP(newsz)) / PGSIZE;
+    uvmunmap(pagetable, PGROUNDUP(newsz), npages, 0);
+  }
+
+  return newsz;
+}
+
+
+
+
+
 // mark a PTE invalid for user access.
 // used by exec for the user stack guard page.
 void
diff --git a/user/user.h b/user/user.h
index 4d398d5..355d38b 100644
--- a/user/user.h
+++ b/user/user.h
@@ -22,7 +22,12 @@ int getpid(void);
 char* sbrk(int);
 int sleep(int);
 int uptime(void);
-
+int create_thread(void (*fcn)(void*) , void* arg , void* stack) ;
+int exit_thread(void) ;
+int join_thread(int) ;
+int yield(int) ;
+int release_sleep(uint8*) ;
+int wakeup(int) ;
 // ulib.c
 int stat(const char*, struct stat*);
 char* strcpy(char*, const char*);
diff --git a/user/usys.pl b/user/usys.pl
index 01e426e..3a84847 100755
--- a/user/usys.pl
+++ b/user/usys.pl
@@ -36,3 +36,9 @@ entry("getpid");
 entry("sbrk");
 entry("sleep");
 entry("uptime");
+entry("create_thread") ;
+entry("exit_thread") ;
+entry("join_thread") ;
+entry("yield") ;
+entry("release_sleep") ;
+entry("wakeup") ;
